include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -186,297 +186,242 @@ class __kernel_name_base
+                __brick(__idx, __rngs...);
+            });
         });
-    });
-    return __future<void>(__event);
-}
+        return __future<void>(__event);
+    }
+};
 
-//------------------------------------------------------------------------
-// parallel_transform_reduce - async pattern
-//------------------------------------------------------------------------
+template <typename _KernelName, typename _Tp, ::std::size_t _Grainsize>
+struct __parallel_reduce_invoker;
 
-template <typename _Tp, ::std::size_t __grainsize = 4, typename _ExecutionPolicy, typename _Up, typename _Cp,
-          typename _Rp, typename... _Ranges>
-oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy,
-                                                             oneapi::dpl::__par_backend_hetero::__future<_Tp>>
-__parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _Rp __brick_reduce, _Ranges&&... __rngs)
+template <typename... _Name, typename _Tp, ::std::size_t _Grainsize>
+struct __parallel_reduce_invoker<__internal::__optional_kernel_name<_Name...>, _Tp, _Grainsize>
 {
-    auto __n = oneapi::dpl::__ranges::__get_first_range_size(__rngs...);
-    assert(__n > 0);
+    template <typename _ExecutionPolicy, typename _Up, typename _Cp,
+          typename _Rp, typename... _Ranges>
+     oneapi::dpl::__par_backend_hetero::__future<_Tp> operator()(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _Rp __brick_reduce, _Ranges&&... __rngs) const
+    {
+        auto __n = oneapi::dpl::__ranges::__get_first_range_size(__rngs...);
+        assert(__n > 0);
 
-    using _Size = decltype(__n);
-    using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
-    using _CustomName = typename _Policy::kernel_name;
-    using _ReduceKernel =
-        oneapi::dpl::__par_backend_hetero::__internal::_KernelName_t<__parallel_reduce_kernel, _CustomName, _Up, _Cp,
-                                                                     _Rp, _Ranges...>;
-
-    sycl::cl_uint __max_compute_units = oneapi::dpl::__internal::__max_compute_units(__exec);
-    ::std::size_t __work_group_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
-    // change __work_group_size according to local memory limit
-    __work_group_size = oneapi::dpl::__internal::__max_local_allocation_size<_ExecutionPolicy, _Tp>(
-        ::std::forward<_ExecutionPolicy>(__exec), __work_group_size);
+        using _Size = decltype(__n);
+
+        sycl::cl_uint __max_compute_units = oneapi::dpl::__internal::__max_compute_units(__exec);
+        ::std::size_t __work_group_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+        // change __work_group_size according to local memory limit
+        __work_group_size = oneapi::dpl::__internal::__max_local_allocation_size<_ExecutionPolicy, _Tp>(
+            ::std::forward<_ExecutionPolicy>(__exec), __work_group_size);
 #if _ONEDPL_COMPILE_KERNEL
```

Comment:
Obviously, you should do re-basing first, before the changes... hope that the merge will be simple.

Reply:
Rebase is done